notebooks/ch17/classification_multiclass.md

```diff
@@ -0,0 +1,312 @@
+
+### One-Vs-Rest Classification
+
+In OvR classification (also known as one-vs-all, or OvA), we decompose a multiclass classification problem into several different binary classification problems. For example, we might observe training data as shown below:
+
+
+```python
+# HIDDEN
+shapes = pd.DataFrame(
+    [[1.3, 3.6, 'triangle'], [1.6, 3.2, 'triangle'], [1.8, 3.8, 'triangle'],
+     [2.0, 1.2, 'square'], [2.2, 1.9, 'square'], [2.6, 1.4, 'square'],
+     [3.2, 2.9, 'circle'], [3.5, 2.2, 'circle'], [3.9, 2.5, 'circle']],
+    columns=['$x_1$', '$x_2$', '$y$']
+)
+```
+
+
+```python
+sns.lmplot('$x_1$', '$x_2$', data=shapes, hue='$y$', markers=['^', 's', 'o'], fit_reg=False)
+plt.xlim(1.0, 4.0)
+plt.ylim(1.0, 4.0);
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_5_0.png)
+
+
+Our goal is to build a multiclass classifier that labels observations as `triangle`, `square`, or `circle` given values for $x_1$ and $x_2$. First, we want to build a binary classifier that predicts observations as `triangle`/not `triangle`:
+
+
+```python
+plot_binary(shapes, 'triangle')
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_7_0.png)
+
+
+Similarly, we build binary classifiers for the remaining classes:
+
+
+```python
+plot_binary(shapes, 'square')
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_9_0.png)
+
+
+
+```python
+plot_binary(shapes, 'circle')
+```
+
+
+![png](classification_multiclass_files/classification_multiclass_10_0.png)
+
+
+We know that the output of the sigmoid function in logistic regression is a probability value from 0 to 1. Then to solve our multiclass classification task, we find the probability of the positive class in each binary classifier and select the class that outputs the highest positive class probability. For example, if we have a new observation with the following values:
+
+$x_1$ | $x_2$
+---   | ---
+3.2   | 2.5
+
+Then our multiclass classifier would fit three binary classifiers to each of the labels of `triangle`, `square`, and `circle`. We extract the positive class probability of each of the three classifiers:
+
+
+```python
+# HIDDEN
+lr_triangle = LogisticRegression(random_state=42)
+lr_triangle.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'triangle')
+proba_triangle = lr_triangle.predict_proba([[3.2, 2.5]])[0][1]
+
+lr_square = LogisticRegression(random_state=42)
+lr_square.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'square')
+proba_square = lr_square.predict_proba([[3.2, 2.5]])[0][1]
+
+lr_circle = LogisticRegression(random_state=42)
+lr_circle.fit(shapes[['$x_1$', '$x_2$']], shapes['$y$'] == 'circle')
+proba_circle = lr_circle.predict_proba([[3.2, 2.5]])[0][1]
+
+print('Triangle:', proba_triangle)
+print('Square:', proba_square)
+print('Circle:', proba_circle)
```

Comment:
Might as well omit the print statements if we're displaying the table right after.

Reply:
Personally I feel it would be beneficial to show only the three `predict_proba` lines of the cell, keep the print statements, and omit the DataFrame, but this works too!

Reply:
I think in this snippet, it's better to introduce just the theory behind OvR (and thus show just the table and not the `predict_proba`). In practice, students probably won't be running `predict_proba` on each class and then selecting the class with the highest output; it might be clearer to the reader to let the only code implementations shown in this section to be the default OvR implementation in LogisticRegression (shown later in the Iris example).